worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,131 @@
+    val connectionProvider: Option[HikariCpPoolingConnectionProvider] = createNewConnectionProvider(finalJdbcURL)
+    logger.info(s"Initialized Db connection pool for ${finalJdbcURL}")
+    // Hikari takes about a second to add connections to the connection pool
+    // We are now going to warm-up connectionPool(with timelimit of 2500ms)
+    connectionProvider.map(warmUpCP)
+  }
+
   val logger = LoggerFactory.getLogger(this.getClass)
   private val dataSource = props.getProperty("org.quartz.jobStore.dataSource")
-  private val provider = if(dataSource != null) {
-    Some(new HikariCpPoolingConnectionProvider(
-      props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".URL"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
-      props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
-    ))
-  } else {
-    logger.info("No job store found in config")
-    None
+  private val originalJdbcURL = props.getProperty("org.quartz.dataSource." + dataSource + ".URL")
+  private val detectIpAddressFailover = props.getProperty("org.quartz.dataSource." + dataSource + ".ipFailover") == "true"
+  // Removes "jdbc:mysql://" prefix and ":{port}..." suffix
+  private val dataSourceHostname = if (originalJdbcURL != null) originalJdbcURL.replace("jdbc:mysql://", "").split(":")(0) else null
+
+  // Needs to be mutable so that the pool can be exchanged during a failover AND kept in memory so that connections are long-lived
+  private var pool: Pool = new Pool(getIP)
+  private val poolLock = new Object()
+
+  def createNewConnectionProvider(finalJdbcURL: String): Option[HikariCpPoolingConnectionProvider] = {
+    if(dataSource != null) {
+      Some(new HikariCpPoolingConnectionProvider(
+        props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
+        finalJdbcURL,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
+      ))
+    } else {
+      logger.info("No job store found in config to get connections")
+      None
+    }
+  }
+
+  /**
+   * HikariCP connection pools don't automatically close when IP addresses for a hostname change. This function returns True, iff at
+   * least one of the following conditions is met:
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *   - causeFailoverEveryConnection is set to "true", which is used for testing failover functionality
```

Comment:
We can turn this method from private to protected so we can override it with a class that extends `ConnectionProvider` Might be the quickest way to do it. I'll give it a shot

Reply:
I think the way to do it would be to inject a DNS resolver object.  In a real run, you would use something that called `java.net.InetAddress.getByName`, but in a test you would use something that returned a different ip address each time (or changed the ip address at a controlled time).